cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/service/DatasetInstanceHandler.java

```diff
@@ -119,6 +128,17 @@ public void getInfo(HttpRequest request, HttpResponder responder, @PathParam("na
       }
       // typeMeta is guaranteed to be non-null now.
       DatasetMeta info = new DatasetMeta(spec, typeMeta, null);
+      if (ownerType != null && ownerId != null) {
+        try {
+          if (ownerType.equals(Id.getType(Id.Program.class))) {
+            usageRegistry.register(Id.Program.fromStrings(ownerId.split("/")), datasetId);
```

Comment:
Transactions don't make it thread-safe. But since you get a new instance of the dataset for each transactions, it should be fine.